js.go

```diff
@@ -0,0 +1,1246 @@
+func (js *js) QueueSubscribe(subj, queue string, cb MsgHandler, opts ...SubOpt) (*Subscription, error) {
+	return js.subscribe(subj, queue, cb, nil, opts)
+}
+
+// Subscribe will create a subscription to the appropriate stream and consumer.
+func (js *js) ChanSubscribe(subj string, ch chan *Msg, opts ...SubOpt) (*Subscription, error) {
+	return js.subscribe(subj, _EMPTY_, nil, ch, opts)
+}
+
+// ApiPaged includes variables used to create paged responses from the JSON API
+type ApiPaged struct {
+	Total  int `json:"total"`
+	Offset int `json:"offset"`
+	Limit  int `json:"limit"`
+}
+
+type streamRequest struct {
+	Subject string `json:"subject,omitempty"`
+}
+
+type JSApiStreamNamesResponse struct {
+	ApiResponse
+	ApiPaged
+	Streams []string `json:"streams"`
+}
+
+func (js *js) subscribe(subj, queue string, cb MsgHandler, ch chan *Msg, opts []SubOpt) (*Subscription, error) {
+	cfg := ConsumerConfig{AckPolicy: ackPolicyNotSet}
+	o := subOpts{cfg: &cfg}
+	if len(opts) > 0 {
+		for _, f := range opts {
+			if err := f(&o); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	isPullMode := o.pull > 0
+	if cb != nil && isPullMode {
+		return nil, ErrPullModeNotAllowed
+	}
+
+	var err error
+	var stream, deliver string
+	var ccfg *ConsumerConfig
+
+	// If we are attaching to an existing consumer.
+	shouldAttach := o.stream != _EMPTY_ && o.consumer != _EMPTY_ || o.cfg.DeliverSubject != _EMPTY_
+	shouldCreate := !shouldAttach
+
+	if js.direct && shouldCreate {
+		return nil, ErrDirectModeRequired
+	}
+
+	if js.direct {
+		if o.cfg.DeliverSubject != _EMPTY_ {
+			deliver = o.cfg.DeliverSubject
+		} else {
+			deliver = NewInbox()
+		}
+	} else if shouldAttach {
+		info, err := js.getConsumerInfo(o.stream, o.consumer)
+		if err != nil {
+			return nil, err
+		}
+
+		ccfg = &info.Config
+		// Make sure this new subject matches or is a subset.
+		if ccfg.FilterSubject != _EMPTY_ && subj != ccfg.FilterSubject {
+			return nil, ErrSubjectMismatch
+		}
+		if ccfg.DeliverSubject != _EMPTY_ {
+			deliver = ccfg.DeliverSubject
+		} else {
+			deliver = NewInbox()
+		}
+	} else {
+		stream, err = js.lookupStreamBySubject(subj)
+		if err != nil {
+			return nil, err
+		}
+		deliver = NewInbox()
+		if !isPullMode {
+			cfg.DeliverSubject = deliver
+		}
+		// Do filtering always, server will clear as needed.
+		cfg.FilterSubject = subj
+	}
+
+	var sub *Subscription
+
+	// Check if we are manual ack.
+	if cb != nil && !o.mack {
+		ocb := cb
+		cb = func(m *Msg) { ocb(m); m.Ack() }
+	}
+
+	sub, err = js.nc.subscribe(deliver, queue, cb, ch, cb == nil)
+	if err != nil {
+		return nil, err
+	}
+	sub.jsi = &jsSub{js: js}
+
+	// If we are creating or updating let's process that request.
+	if shouldCreate {
+		// If not set default to ack explicit.
+		if cfg.AckPolicy == ackPolicyNotSet {
+			cfg.AckPolicy = AckExplicit
+		}
+		// If we have acks at all and the MaxAckPending is not set go ahead
+		// and set to the internal max.
```

Comment:
In NATS streaming IIRC we found better performance and stability with a much lower number - e.g. 32 or 64.  In java, the default pending limit is 64k - imo that'd put a lot of memory pressure on the client.  /cc @kozlovic

Reply:
Right. The default in NATS Streaming is 1024. Having this number high will cause the server to load/resend a lot of messages when not ack'ed in a timely fashion, or when the server has not yet detected the subject interest is gone (right now it is local, so not an issue, but may become with clustering in place).

Reply:
We have two limits, one for number of messages and one for byte size pending which is still applicable.